src/libraries/System.Net.Http/tests/FunctionalTests/Socks/LoopbackSocksServer.cs

```diff
@@ -0,0 +1,365 @@
+                }
+                catch (Exception ex)
+                {
+                    EventSourceTestLogging.Log.TestAncillaryError(ex);
+                }
+
+                _serverStopped.Set();
+            });
+        }
+
+        private async Task ProcessConnection(Socket s)
+        {
+            Interlocked.Increment(ref _connections);
+
+            using (var ns = new NetworkStream(s, ownsSocket: true))
+            {
+                await ProcessRequest(s, ns).ConfigureAwait(false);
+            }
+        }
+
+        private async Task ProcessRequest(Socket clientSocket, NetworkStream ns)
+        {
+            int version = await ns.ReadByteAsync().ConfigureAwait(false);
+
+            await (version switch
+            {
+                4 => ProcessSocks4Request(clientSocket, ns),
+                5 => ProcessSocks5Request(clientSocket, ns),
+                -1 => throw new Exception("Early EOF"),
+                _ => throw new Exception("Bad request version")
+            }).ConfigureAwait(false);
+        }
+
+        private async Task ProcessSocks4Request(Socket clientSocket, NetworkStream ns)
+        {
+            byte[] buffer = new byte[7];
+            await ReadToFillAsync(ns, buffer).ConfigureAwait(false);
+
+            if (buffer[0] != 1)
+                throw new Exception("Only CONNECT is supported.");
+
+            int port = (buffer[1] << 8) + buffer[2];
+            // formats ip into string to ensure we get the correct order
+            string remoteHost = $"{buffer[3]}.{buffer[4]}.{buffer[5]}.{buffer[6]}";
+
+            byte[] usernameBuffer = new byte[1024];
+            int usernameBytes = 0;
+            while (true)
+            {
+                int usernameByte = await ns.ReadByteAsync().ConfigureAwait(false);
+                if (usernameByte == 0)
+                    break;
+                if (usernameByte == -1)
+                    throw new Exception("Early EOF");
+                usernameBuffer[usernameBytes++] = (byte)usernameByte;
+            }
+
+            if (remoteHost.StartsWith("0.0.0") && remoteHost != "0.0.0.0")
+            {
+                byte[] hostBuffer = new byte[1024];
+                int hostnameBytes = 0;
+
+                while (true)
+                {
+                    int b = await ns.ReadByteAsync().ConfigureAwait(false);
+                    if (b == -1)
+                        throw new Exception("Early EOF");
+                    if (b == 0)
+                        break;
+
+                    hostBuffer[hostnameBytes++] = (byte)b;
+                }
+
+                remoteHost = Encoding.UTF8.GetString(hostBuffer.AsSpan(0, hostnameBytes));
+            }
+
+            if (_username != null)
+            {
+                string username = Encoding.UTF8.GetString(usernameBuffer.AsSpan(0, usernameBytes));
+                if (username != _username)
+                {
+                    ns.WriteByte(4);
```

Comment:
no need for async?

Reply:
It's a test only server. I just want it to be simple.

Reply:
There isn't a helper to write 1 byte asynchronously. So I'm not willing to bother with it.